jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp

```diff
@@ -0,0 +1,369 @@
+
+  bool
+  Apply(PointerObjectSet& set);
+};
+
+/**
+ * A constraint of the form:
+ * P(loaded) is a superset of P(x) for all x in P(pointer)
+ * Example of application is a load, e.g. when loaded = *pointer
+ */
+class SupersetOfAllPointeesConstraint final
+{
+  PointerObject::Index Loaded_;
+  PointerObject::Index Pointer_;
+
+public:
+  SupersetOfAllPointeesConstraint(PointerObject::Index loaded, PointerObject::Index pointer)
+                                 : Loaded_(loaded), Pointer_(pointer) {}
+
+  bool
+  Apply(PointerObjectSet& set);
+};
+
+/**
+ * A class for adding and applying constraints to the points-to-sets of the PointerObjectSet.
+ * Unlike the set modification methods on PointerObjectSet, constraints can be added in any order, with the same result.
+ * Use Solve() to calculate the final points-to-sets.
+ *
+ * Constraints on the special nodes, external and escaped, are built in.
+ */
+class PointerObjectConstraintSet final
+{
+public:
+  using ConstraintVariant = std::variant<
+          SupersetConstraint,
+          AllPointeesPointToSupersetConstraint,
+          SupersetOfAllPointeesConstraint>;
+
+  explicit PointerObjectConstraintSet(PointerObjectSet& set) : Set_(set) {}
+
+  PointerObjectConstraintSet(const PointerObjectConstraintSet& other) = delete;
+
+  PointerObjectConstraintSet(PointerObjectConstraintSet&& other) = delete;
+
+  PointerObjectConstraintSet&
+  operator =(const PointerObjectConstraintSet& other) = delete;
+
+  PointerObjectConstraintSet&
+  operator =(PointerObjectConstraintSet&& other) = delete;
+
+  /**
+   * The simplest constraint, on the form: pointee in P(pointer)
+   * @param pointer the PointerObject that should have the pointee in its points-to set
+   * @param pointee the PointerObject that should be in the points-to-set
+   */
+  void
+  AddPointerPointeeConstraint(PointerObject::Index pointer, PointerObject::Index pointee);
+
+  /**
+   * Adds a constraint making \p pointer flagged as pointing to external
+   * @param pointer the PointerObject that should be marked as pointing to external
+   */
+  void
+  AddPointsToExternalConstraint(PointerObject::Index pointer);
+
+  /**
+   * Ensures that any PointerObject in P(registerIndex) will be marked as escaped.
+   * @param registerIndex the register whose content leaves the module, thus exposing any memory it may point to
+   */
+  void
+  AddRegisterContentEscapedConstraint(PointerObject::Index registerIndex);
+
+  /**
+   * Generic add function for all struct based constraints
+   * @param c an instance of a constraint struct, passed as a ConstraintVariant
+   */
+  void
+  AddConstraint(ConstraintVariant c);
+
+  /**
+   * Iterates over and applies constraints until the all points-to-sets satisfy them
+   */
+  void
+  Solve();
+
+private:
+
+  /**
+   * Makes sure any PointerObject marked as escaped, also makes all its pointees get the HasEscaped flag set.
```

Comment:
"Ensures that the escaped flag is set for all pointees of any pointer object that is marked as escaped, "